lib/puma/json.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+module Puma
+
+  # Puma deliberately avoids the use of the json gem and instead performs JSON
+  # serialization without any external dependencies. In a puma cluster, loading
+  # any gem into the puma master process means that operators cannot use a
+  # phased restart to upgrade their application if the new version of that
+  # application uses a different version of that gem. The json gem in
+  # particular is additionally problematic because it leverages native
+  # extensions. If the puma master process relies on a gem with native
+  # extensions and operators remove gems from disk related to old releases,
+  # subsequent phased restarts can fail.
+  #
+  # The implementation of JSON serialization in this module is not designed to
+  # be particularly full-featured or fast. It just has to handle the few places
+  # where Puma relies on JSON serialization internally.
+
+  module JSON
+    class SerializationError < StandardError; end
+
+    class << self
+      def generate(value)
+        case value
+        when Array, Hash
+          parts = []
```

Comment:
Rather than an array, should it use [StringIO](https://ruby-doc.org/stdlib-2.5.1/libdoc/stringio/rdoc/StringIO.html)? Not sure if there is a (performance) benefit for one or the other.

Reply:
I considered using `stringio` since it seems like a nice way to construct a string incrementally without an explosion in extra space. I went with just a normal `Array` because even `stringio` isn't in Ruby Core and, just like `json`, users can technically change the version of `stringio` they're using in a phased restart. That said `puma` _already_ uses `stringio` internally, so I could go either way on this.

I can't imagine the performance improvement would be too significant, especially since again, this module isn't really on a hot path.

Reply:
The reason I'd switch to `stringio` is if it would improve readability/maintainability. I'll play with it and see what it looks like.

Reply:
Switch to StringIO in 513aab1ea343bbc3c9e2adcc3acb8645aff64c74.

I ended up doing this because I think using the `Array` approach I was doing before stood out as unusual after thinking about it. If someone was trying to build a String incrementally while keep allocations to a minimum, a reader of the code would expect the author to have used `stringio`. Deviating from that expectation would suggest that the distinction was significant, but in this case, it isn't.